crates/sui-types/src/crypto.rs

```diff
@@ -33,21 +35,129 @@ use crate::error::{SuiError, SuiResult};
 use crate::sui_serde::{Base64, Readable, SuiBitmap};
 pub use enum_dispatch::enum_dispatch;
 
-// Comment the one you want to use
-
 // Authority Objects
 pub type AuthorityKeyPair = Ed25519KeyPair;
 pub type AuthorityPrivateKey = Ed25519PrivateKey;
 pub type AuthorityPublicKey = Ed25519PublicKey;
 pub type AuthoritySignature = Ed25519Signature;
 pub type AggregateAuthoritySignature = Ed25519AggregateSignature;
 
-// Account Objects
+// TODO(joyqvq): prefix these types with Default, DefaultAccountKeyPair etc
 pub type AccountKeyPair = Ed25519KeyPair;
 pub type AccountPublicKey = Ed25519PublicKey;
 pub type AccountPrivateKey = Ed25519PrivateKey;
 pub type AccountSignature = Ed25519Signature;
 
+#[enum_dispatch]
+#[derive(Debug)]
+pub enum SuiKeyPair {
+    Ed25519KeyPair,
+    Secp256k1KeyPair,
+}
+
+#[derive(Debug)]
+pub enum PublicKey<'a> {
+    Ed25519KeyPair(&'a Ed25519PublicKey),
+    Secp256k1KeyPair(&'a Secp256k1PublicKey),
+}
+
+impl SuiKeyPair {
+    pub fn public(&self) -> PublicKey<'_> {
+        match self {
+            SuiKeyPair::Ed25519KeyPair(kp) => PublicKey::Ed25519KeyPair(kp.public()),
+            SuiKeyPair::Secp256k1KeyPair(kp) => PublicKey::Secp256k1KeyPair(kp.public()),
+        }
+    }
+}
+
+impl signature::Signer<Signature> for SuiKeyPair {
+    fn try_sign(&self, msg: &[u8]) -> Result<Signature, signature::Error> {
+        match self {
+            SuiKeyPair::Ed25519KeyPair(kp) => kp.try_sign(msg),
+            SuiKeyPair::Secp256k1KeyPair(kp) => kp.try_sign(msg),
+        }
+    }
+}
+
+impl FromStr for SuiKeyPair {
+    type Err = eyre::Report;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let kp = Self::decode_base64(s).map_err(|e| eyre::eyre!("{}", e.to_string()))?;
+        Ok(kp)
+    }
+}
+
+impl EncodeDecodeBase64 for SuiKeyPair {
+    fn encode_base64(&self) -> String {
+        let mut bytes: Vec<u8> = Vec::new();
+        match self {
+            SuiKeyPair::Ed25519KeyPair(kp) => {
```

Comment:
You should have an implementation of `ToFromBytes` for each of `Ed25519Keypair` and `Secp256k1KeyPair`, because I believe you use it in `from_bytes` above.

Why not simply use `as_bytes`on the appropriate variant, which would absolve you from mucking about with public and private key compoents (at the risk of misordering them!) ?

Reply:
the issue with ToFromBytes impl is that we are missing AsRef both key pairs (unless we want to add it to the KeyPairTraits in narwhal?)